### PR TITLE
TMDM-13611 Investigate the performance degregation for list ids

### PR DESCRIPTION
**Pull request by hw-zhu**
https://jira.talendforge.org/browse/TMDM-13611
**What is the current behavior?** (You should also link to an open issue here)
improvement point: Add a new functionality "count" in task TMDM-13266, so if hasn't field "count" in return DataRecord list, that will cause a exception.the code snippets cause a performance degregation.

**What is the new behavior?**
Add a hasField statement to check if the field exist or not, avoid the exception.
1. Reduce the deep of method stack.
2. Using code conditional judgment instead of catch exception.
3. Reduce the number of list iterator to improve performance.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java [hunk 1]
```diff
@@ -90,8 +97,12 @@ public Void visit(ContainedTypeFieldMetadata containedField) {
                             if (delegator.hide(containedField)) {
                                 return null;
                             }
+                            String fieldName = field.getName();
+                            if (ignoreCase) {
+                                fieldName = fieldName.toLowerCase();
+                            }
                             try {
-                                writer.key(getFieldName(containedField.getName()));
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: The old name is `containedField`, the new one is `field`, is that Ok?

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java [hunk 2]
```diff
@@ -124,6 +135,38 @@ public Void visit(ReferenceFieldMetadata referenceField) {
...
+            DataRecord record = iterator.next();
+            write(record, jsonWriter);
+            if (iterator.hasNext() && MediaType.APPLICATION_JSON_TYPE.equals(mediaType)) {
+                output.write(",".getBytes()); //$NON-NLS-1$
+            }
+            writer.flush();
+        }
+    }
+
+    public void write(DataRecord record, JSONWriter jsonWriter) throws IOException {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: Is this a public or private function?

File: org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java [hunk 3]
```diff
@@ -194,8 +194,18 @@ public Object get(String fieldName) {
...
         Object currentValue = null;
+        boolean hasField = false;
         while (tokenizer.hasMoreTokens()) {
             String currentPathElement = tokenizer.nextToken();
+            if(!current.getType().hasField(currentPathElement)) {
+                continue;
+            } else {
+                hasField = true;
+            }
+            if (!hasField) {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java]: Can this be possibly called in the while loop or out of the while loop?

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java (hunk 1 above)
Comment (hw-zhu) [reply org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: done

File: org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java (hunk 3 above)
Comment (hw-zhu) [reply org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java]: Each of item need to has a if condition to decide the current loop status.

File: org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java [hunk 4]
```diff
@@ -194,8 +194,18 @@ public Object get(String fieldName) {
         StringTokenizer tokenizer = new StringTokenizer(fieldName, "/"); //$NON-NLS-1$
         DataRecord current = this;
         Object currentValue = null;
+        boolean hasField = false;
         while (tokenizer.hasMoreTokens()) {
             String currentPathElement = tokenizer.nextToken();
+            if(!current.getType().hasField(currentPathElement)) {
```
Comment (lpw82gf) [inline org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java]: Format

Comment (hw-zhu) [reply org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java]: done

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java [hunk 5]
```diff
@@ -15,8 +15,11 @@
 import java.io.OutputStreamWriter;
 import java.io.Writer;
 import java.nio.charset.Charset;
+import java.util.Iterator;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: Need to add `//$NON-NLS-1$` for all string constant

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java [hunk 6]
```diff
@@ -124,6 +135,38 @@ public Void visit(ReferenceFieldMetadata referenceField) {
         writer.endObject();
     }
 
+    @Override
+    public void write(StorageResults recordList, OutputStream output, MediaType mediaType) throws IOException {
+        Writer writer = new OutputStreamWriter(output, Charset.forName("UTF-8"));
+        for (Iterator<DataRecord> iterator = recordList.iterator(); iterator.hasNext();) {
+            JSONWriter jsonWriter = new JSONWriter(writer);
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: Can `jsonWriter` move out of `for` loop?

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java [hunk 7]
```diff
@@ -124,6 +135,38 @@ public Void visit(ReferenceFieldMetadata referenceField) {
...
+            DataRecord record = iterator.next();
+            write(record, jsonWriter);
+            if (iterator.hasNext() && MediaType.APPLICATION_JSON_TYPE.equals(mediaType)) {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: May not need `MediaType.APPLICATION_JSON_TYPE.equals(mediaType)`  here.

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java [hunk 8]
```diff
@@ -124,6 +135,38 @@ public Void visit(ReferenceFieldMetadata referenceField) {
...
+                output.write(",".getBytes()); //$NON-NLS-1$
+            }
+            writer.flush();
+        }
+    }
+
+    private void write(DataRecord record, JSONWriter jsonWriter) throws IOException {
+        String fieldName = record.getType().getName();
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: `fieldName` to `typeName` ?

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java [hunk 9]
```diff
@@ -132,8 +175,12 @@ public void write(DataRecord record, OutputStream output) throws IOException {
     @Override
     public void write(DataRecord record, Writer writer) throws IOException {
         JSONWriter jsonWriter = new JSONWriter(writer);
+        String fieldName = record.getType().getName();
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: Consider to have a util method to deal with the `ignoreCase` logic for all the places.

File: org.talend.mdm.core/src/com/amalto/core/storage/record/ViewSearchResultsWriter.java [hunk 10]
```diff
@@ -15,23 +15,40 @@
...
+import com.amalto.core.storage.SecuredStorage;
 import com.amalto.core.storage.StorageMetadataUtils;
+import com.amalto.core.storage.StorageResults;
 
 public class ViewSearchResultsWriter implements DataRecordWriter {
 
     private SecuredStorage.UserDelegator delegator = SecuredStorage.UNSECURED;
 
+    @Override
+    public void write(StorageResults recordList, OutputStream output, MediaType mediaType) throws IOException {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/storage/record/ViewSearchResultsWriter.java]: For the all implementations almost get same code for this method, consider 
- do we need to implement it in all classes
- can we share the code, seems the logic is the same for most of them

File: org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java (hunk 6 above)
Comment (hw-zhu) [reply org.talend.mdm.core/src/com/amalto/core/storage/record/DataRecordJSONWriter.java]: need to create a new jsonwriter object for each of loop

**Pull request merged**